### PR TITLE
feat: allow developers to retry an unsuccessful request

### PR DESCRIPTION
**Pull request by lucasaraujo**
## Description

Allow developers to retry an unsuccessful request.
Added a retry block to `ServerDrivenState.error`.
BeagleDemo updated to present an error view when the screen load fails, the error view has a button to retry the request.

## Related Issues

Part of #585

## Tests

Updated test to use the retry.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [DCO].
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] Yes, this is a breaking change.

BREAKING CHANGE: added new associated valued of type `BeagleRetry` to enum case `ServerDrivenState.error` 

[issue database]: https://github.com/ZupIT/beagle/issues
[DCO]: https://developercertificate.org/

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: iOS/Sources/Beagle/Sources/Action/Action.swift [hunk 1]
```diff
@@ -25,3 +25,12 @@ extension UnknownAction: Action {
         controller.dependencies.logger.log(Log.decode(.decodingError(type: "error trying to execute unknown action")))
     }
 }
+
+extension Action {
+    public func retryClosure(controller: BeagleController, sender: Any) -> BeagleRetry {
```
Comment (theffc) [inline iOS/Sources/Beagle/Sources/Action/Action.swift:30]: Maybe a name like `closureToRetrySameAction` would make the actual behavior clearer

File: iOS/Sources/Beagle/Sources/Renderer/BeagleNavigationController.swift [hunk 2]
```diff
@@ -33,6 +35,16 @@ extension ServerDrivenState {
...
+    /// This method is the entry point to handle the screen state changes.
+    /// The state tells when the screen is loading or any error happend.
+    /// The default implemetation shows an `ActivityIndicator` when the screen is
+    /// loading and does nothing when error happens, override this method to handle
+    /// the erros properly. When overriding, the `super` implementation should be called
+    /// to preserve the loading or you will have to handle the loading.
+    ///
+    /// - Parameters:
+    ///   - state: the new state
+    ///   - screenController: controller that triggered the state change
```
Comment (theffc) [inline iOS/Sources/Beagle/Sources/Renderer/BeagleNavigationController.swift:47]:
```suggestion
    /// This method is the entry point to handle screen state changes.
    /// The default implemetation shows an `ActivityIndicator` when screen is
    /// loading and does nothing when error happens; override this method to handle
    /// errors properly.
    /// When overriding, if you want to preserve loading behavior, `super` implementation should be called,
    /// or you can customize loading behavior yourself.
    ///
    /// - Parameters:
    ///   - state: new state that tells if screen is loading or any error happened
    ///   - screenController: controller that triggered the state change
```

File: iOS/Sources/Beagle/Sources/Action/Action.swift (hunk 1 above)
Comment (theffc) [inline iOS/Sources/Beagle/Sources/Action/Action.swift:30]: Maybe a name like `closureToRetrySameAction` makes its behavior clearer

File: iOS/Sources/Beagle/Sources/Components/Lazy+Renderable/LazyComponent+Renderable.swift [hunk 3]
```diff
@@ -33,10 +33,20 @@ extension LazyComponent: ServerDrivenComponent {
...
-                renderer.controller.serverDrivenState = .error(.lazyLoad(error))
+                renderer.controller.serverDrivenState = .error(
+                    .lazyLoad(error),
+                    self.retryClosure(initialState: view, renderer: renderer)
+                )
             }
         }
     }
+    
+    public func retryClosure(initialState view: UIView, renderer: BeagleRenderer) -> BeagleRetry {
```
Comment (theffc) [inline iOS/Sources/Beagle/Sources/Components/Lazy+Renderable/LazyComponent+Renderable.swift:44]: should this be public? I don't think we actually need to expose this to our users

Comment (lucasaraujo) [reply iOS/Sources/Beagle/Sources/Components/Lazy+Renderable/LazyComponent+Renderable.swift]: public removed =]

**Pull request merged**